### PR TITLE
feat: implemented reward redistribution and implicit reward redistribution

### PR DESCRIPTION
**Pull request by maharajamihir**

### CODE REVIEW

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -105,6 +106,14 @@ def compute_gae_advantage_return(
         advantages = torch.stack(advantages_reversed[::-1], dim=1)
 
         returns = advantages + values
+
+        # implicit reward redistribution
+        if redistribute_reward_implicit:
+            # Scale the advantages by (T-t)/T
+            scaling_factors = torch.arange(gen_len, 0, -1, dtype=torch.float32) / gen_len
```
Comment (emergenz) [inline verl/trainer/ppo/core_algos.py]: should get the dtype from the config?

Comment (emergenz) [inline verl/trainer/ppo/core_algos.py]: We must divide by the number of valid tokens in the episode, not `MAX_GEN_LEN`

File: verl/workers/reward_manager/naive.py [hunk 2]
```diff
@@ -83,7 +83,12 @@ def __call__(self, data: DataProto, return_dict=False):
             else:
                 reward = score
 
-            reward_tensor[i, valid_response_length - 1] = reward
+            if data_item.meta_info["redistribute_reward"]:
```
Comment (emergenz) [inline verl/workers/reward_manager/naive.py:86]: Do we exclusively use the naive reward manager?

**Review by emergenz — CHANGES_REQUESTED** [review]
Implicit reward redistribution currently contains a bug

File: verl/workers/reward_manager/naive.py (hunk 2 above)
Comment (emergenz) [reply verl/workers/reward_manager/naive.py:86]: Also implement RR for the batch reward manager

File: verl/trainer/ppo/core_algos.py [hunk 3]
```diff
@@ -105,6 +106,17 @@ def compute_gae_advantage_return(
         advantages = torch.stack(advantages_reversed[::-1], dim=1)
 
         returns = advantages + values
+
+        # Scale the advantages by (T-t)/T
+        if redistribute_reward_implicit: 
+            # Compute per‐example length T from the response mask
+            lengths = response_mask.sum(dim=1, keepdim=True).to(advantages)
```
Comment (emergenz) [inline verl/trainer/ppo/core_algos.py:113]: Is the `.to` necessary?

Comment (emergenz) [reply verl/trainer/ppo/core_algos.py:113]: Remove the `.to`

Comment (emergenz) [inline verl/trainer/ppo/core_algos.py:113]:
```suggestion
            lengths = response_mask.sum(dim=1, keepdim=True)
```

File: verl/trainer/ppo/core_algos.py [hunk 4]
```diff
@@ -105,6 +106,17 @@ def compute_gae_advantage_return(
...
+            # Build a [1, T_max] tensor of time indices
+            time_idx = torch.arange(gen_len, device=advantages.device, dtype=advantages.dtype).unsqueeze(0)
```
Comment (emergenz) [inline verl/trainer/ppo/core_algos.py:115]:
```suggestion
            time_idx = torch.arange(gen_len).unsqueeze(0)
```

File: verl/workers/reward_manager/naive.py (hunk 2 above)
Comment (maharajamihir) [reply verl/workers/reward_manager/naive.py:86]: done
